kernel/kernel-api/src/main/java/io/delta/kernel/types/StructField.java

```diff
@@ -79,9 +82,37 @@ public DataType getDataType() {
 
   /** @return the metadata for this field */
   public FieldMetadata getMetadata() {
+    fetchCollationMetadata();
```

Comment:
I am wondering if this can be handled simply by adding the code in the `StructField` constructor. It can go through the type and figure out if it needs collation data to be stored in its metadata. If yes, just add them there. Given StructField is immutable, we don't need to do dynamic computation of the collations for nested fields like here which is prone to bugs.